Duplicati/Library/Backend/Telegram/TelegramBackend.cs

```diff
@@ -0,0 +1,512 @@
+                    }
+                },
+                $"{nameof(Get)}");
+        }
+
+        public Task PutAsync(string remotename, string filename, CancellationToken cancelToken)
+        {
+            using (var fs = File.OpenRead(filename))
+            {
+                PutAsync(remotename, fs, cancelToken).GetAwaiter().GetResult();
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public void Get(string remotename, string filename)
+        {
+            using (var fs = new FileStream(filename, FileMode.Create, FileAccess.Write))
+            {
+                Get(remotename, fs);
+            }
+        }
+
+        public void Delete(string remotename)
+        {
+            SafeExecute(() =>
+                {
+                    var channel = GetChannel();
+                    var fileInfo = ListChannelFileInfos().FirstOrDefault(fi => fi.Name == remotename);
+                    if (fileInfo == null)
+                    {
+                        return;
+                    }
+
+                    var request = new TLRequestDeleteMessages
+                    {
+                        Channel = new TLInputChannel
+                        {
+                            ChannelId = channel.Id,
+                            AccessHash = channel.AccessHash.Value
+                        },
+                        Id = new TLVector<int> {fileInfo.MessageId}
+                    };
+
+                    EnsureConnected();
+                    m_telegramClient.SendRequestAsync<TLAffectedMessages>(request).GetAwaiter().GetResult();
+                },
+                $"{nameof(Delete)}({remotename})");
+        }
+
+        public List<ChannelFileInfo> ListChannelFileInfos()
+        {
+            var channel = GetChannel();
+
+            var inputPeerChannel = new TLInputPeerChannel {ChannelId = channel.Id, AccessHash = channel.AccessHash.Value};
+            var result = new List<ChannelFileInfo>();
+            var cMinId = (int?)0;
+
+            while (cMinId != null)
+            {
+                var newCMinId = RetrieveMessages(inputPeerChannel, result, cMinId.Value);
+                if (newCMinId == cMinId)
+                {
+                    break;
+                }
+
+                cMinId = newCMinId;
+            }
+
+            result = result.Distinct().ToList();
+            return result;
+        }
+
+        private int? RetrieveMessages(TLInputPeerChannel inputPeerChannel, List<ChannelFileInfo> result, int maxDate)
+        {
+            EnsureConnected();
+            var absHistory = m_telegramClient.GetHistoryAsync(inputPeerChannel, offsetDate: maxDate).GetAwaiter().GetResult();
+            var history = ((TLChannelMessages)absHistory).Messages.OfType<TLMessage>();
+            var minDate = (int?)null;
+
+            foreach (var msg in history)
+            {
+                minDate = minDate == null || minDate > msg.Id ? msg.Date : minDate;
```

Comment:
This looks strange, so I just want to confirm.  Do we really want to compare `minDate > msg.Id`?

Reply:
I have refactored the expression, so it is more clear

Reply:
Thanks.  Sorry, I'm not familiar with the Telegram data model, but I just want double check again.  Should this really be `minDate < msg.Id` instead of `minDate < msg.Date`?

Reply:
You are correct there was a bug there. I refactored the code again. I didn't notice it, sorry :)

Reply:
It was working because the messages were coming in the correct order, which we cannot lean on. Fixed, thank you